### DISCUSSION TITLE
Question re interleaving

### DISCUSSION THREAD

**Discussion by omidantilong** [mdx-js/mdx#2038] [Q&A] [answered]
Hello (again) 🙂 I have another question about interleaving. Specifically about content being wrapped in `<p>` tags. 

https://mdxjs.com/docs/what-is-mdx/#interleaving

As I understand it, those docs say are saying that the determination on whether to wrap in `<p>` tags is based on whether or not the content is on the same line? 

Unfortunately most editors will automatically wrap long lines inside block level elements so this is not really practical. I had always thought markdown (and commonmark, so by extension MDX) did not do anything to content inside block elements embedded in markdown. 

The only way I have found to get around this is to wrap the entire `p` tag in curly braces but as the whole purpose of MD is make editing easier, that's not really suitable either. I understand this probably isn't a bug as it's stated in the docs that this is what happens, but the rule of "is it on the same line" seems a little too inflexible given the nature of modern IDEs.

I have looked through remark/rehype plugins for something that might allow me to do this, with no luck so far. Maybe a nice solution would be something in the MDX global options to disable parsing inside embedded HTML? This would also stop the hydration error when using Next.JS as you can end up with multiple nested &lt;p&gt; tags if you need to write the tag out manually - which is perfectly valid such as when class names are required.

Thanks again! :)

<img width="854" alt="Screenshot 2022-05-15 at 01 51 18" src="https://user-images.githubusercontent.com/1922098/168452802-1c157b32-0075-4e53-8574-3f218826bf0d.png">

**Comment by wooorm** [comment, marked as answer]
Hi!

You need to use `` `backticks` `` in markdown on GitHub around HTML. Otherwise they’re treated as actual HTML.

Your understanding of those docs is correct. Your understanding of markdown and HTML blocks is also correct. MDX, as explained in the docs, deviates from that. Our users wanted interleaving.

In your example, you are not using interleaving, and I believe that this fact, *not using interleaving*, makes the interleaving awkward.

```
<div>
  This is already a paragraph. You must not wrap it in `<p>`s.
</div>
```

You can style that `div` whatever way you want.
I understand the theoretical need to style `<p>`s, but believe that in most cases you can instead style a `<div>` around it, which is perhaps better, because the content is not touched, only wrapping components without meaning, receive style.

There are several similar discussions. See also: https://github.com/mdx-js/mdx/issues/1798 and https://github.com/mdx-js/mdx/issues/1946.

You can also configure your editor to not automatically wrap MDX.

**Reply by omidantilong** [reply]
Hey, thanks for the quick response 🙂 I take your point about styling the container but the problem is that no matter what content is inside the `<div>`, it gets wrapped in a `<p>` if the content automatically wraps onto multiple lines. This happens with any element (see the screenshots below - the generated markup contains a `<h1>` with a `<p>`) 

Contents of the .mdx file:
<img width="802" alt="Screenshot 2022-05-15 at 12 35 45" src="https://user-images.githubusercontent.com/1922098/168471336-c12823fd-14c0-49c6-ad68-808c97636ce1.png">

Generated markup with extra p:
<img width="438" alt="Screenshot 2022-05-15 at 12 35 40" src="https://user-images.githubusercontent.com/1922098/168471338-37addcb6-ee16-42ba-ba20-d848430e09a0.png">

I think realistically, there are always going to be times when some child element needs a specific class, however good our intentions are. As you mentioned, as a workaroundI have currently configured the editor+prettier to not wrap MDX and do its own soft wrap instead, which makes scanning files harder so it's not ideal. I guess it also reduces portability because that vscode setting would be editor specific.

This can also cause problems with `<div>short paragraphs</div>` though because they will be formatted back onto one line and the interleaving kicks in. In the screenshot below the contents of the div will be formatted back onto one line as soon you hit save, unless you configure the editor to completely ignore mdx files. Which feels like a bit of a nuclear option!  
<img width="429" alt="Screenshot 2022-05-15 at 12 58 19" src="https://user-images.githubusercontent.com/1922098/168471579-b3f9f4f0-d391-4e87-9d12-e8e50fe3fb44.png">

Cheers :)

**Reply by wooorm** [reply]
> but the problem is

I wonder how often that’s a problem.
I instead think it’s not frequently a problem when embracing interleaving: that markdown is allowed in JSX.
And for the cases where it’s needed, that the workarounds are okay:

```
<p id="stuff">Lorem Ipsum</p>

<p id="stuff">Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do
eiusmod tempor incididunt ut labore et dolore magna aliqua.</p>

{
  <p id="stuff">
    Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do
    eiusmod tempor incididunt ut labore et dolore magna aliqua.
  </p>
}
```

E.g., for your `h1` example, I don’t practically see headings being 80 characters. That’s already hard for users to understand and bad for SEO.

And if the second example doesn’t work with your editor, you can configure your editor.

And if you have an automatic formatter that does these things to MDX, such as prettier, then that’s a bug there.

**Reply by omidantilong** [reply]
Sorry - yes you are right of course, a h1 was a bad example 😄 but you get the same results with any element. 

As for "how often" - well, however often someone writes a long sentence. Or perhaps needs to an id (as in your example) which may contain a particularly long anchor. Or, uses some utility classes in an html element in their mdx files. Or uses a custom component (another of MDX's selling points) which is meant to just take raw text. Or any combination of those!

The real kicker is that it breaks Next's hydration. As I said the determination based on whether or not something is on one line seems a bit too opinionated - at the least a way to disable that so the commonmark standard is observed would make things a lot easier.

I know the workarounds are there such as wrapping in curly braces but the point of markdown (and by extension mdx) is to make authoring simple and predictable, no? 

Thanks again for your replies :) 

<img width="977" alt="Screenshot 2022-05-15 at 17 14 06" src="https://user-images.githubusercontent.com/1922098/168484392-f90a165d-01af-424e-89c1-edd9464ded39.png">

**Reply by wooorm** [reply]
There needs to be a rule. People want interleaving. Your suggested rule is the previous behavior which was not intuitive.

**Reply by wooorm** [reply]
I don’t believe the kicker isn’t hydration. The kicker is there needs to be a rule, and things break if you don’t adhere to the rule. Whether it’s semantic wise, or because Next.js shows it.

As for “the point”, there are different points, and there are trade offs between them. This is intentional

**Reply by omidantilong** [reply]
I understand that - I wasn't aware this was the previous behaviour, did it change between v1->v2? 

My point about "the point" was that MDX flagship feature is that it supports JSX-in-markdown and this breaks existing "rules" in JSX - or at least patterns that people are used to regarding code formatting. Either way, it is what it is so thanks for explaining 👍 

One final question - could this behaviour be overridden with another remark/rehype plugin? I'm struggling to find out where in the pipeline the extra wrapping occurs. I'm interested to see how it works out of curiosity as much as anything.

**Reply by ChristianMurphy** [reply]
You may find the the explanation of the design choice in https://github.com/orgs/mdx-js/discussions/1603 of interest, and in particular the ideas on how unwrapping could be done in https://github.com/orgs/mdx-js/discussions/1603#discussioncomment-1091977.

**Reply by omidantilong** [reply]
@ChristianMurphy thanks for those links, that discussion is interesting. I know some markdown implementations support a `markdown=0` attribute on a wrapping element - I wonder if that's something MDX could consider for these situations?

Cheers for pointing on the right path for the plugin stuff :)

**Reply by omidantilong** [reply]
I have a working plugin for this now. I don't know if this is the most efficient way but it does exactly what I want. 

```
export default function remarkUnwrapTags() {
 return (tree) => {
   visit(tree, 'paragraph', (node, index, parent) => {
     if(node.children) {
       node.children.forEach((child) => {
         if(child.value) {
           child.value = child.value.replaceAll("\n", '').replaceAll("\r"," ").toString()
         }
       })
     }

     if(parent.type==='mdxJsxFlowElement') {
       parent.children = node.children
     }
    
   })
 }
}
```

**Reply by ChristianMurphy** [reply]
>  I know some markdown implementations support a `markdown=0` attribute on a wrapping element

Oh? Which implementation is that?
None of the one's in Babelmark appear to support it: https://babelmark.github.io/?text=%3Cp+markdown%3D0%3E%0A%0A%23+heading%0A%0A%3C%2Fp%3E%0A%0A%3Cp%3E%0A%0A%23+heading%0A%0A%3C%2Fp%3E

> I wonder if that's something MDX could consider for these situations?

It could be added via plugin.
I don't see this functionality as something which would be added directly to core.

> I have a working plugin for this now. I don't know if this is the most efficient way but it does exactly what I want.

I'm not quite sure I follow the part where all of the newlines need to be replaced.
But if it work, it works 🙂

**Reply by omidantilong** [reply]
> Oh? Which implementation is that?

There is some discussion on this very topic over on commonmark's talk page: https://talk.commonmark.org/t/is-markdown-0-only-used-by-kramdown/1995 - it does appear to be a limited set, and in the babelmark test you linked to the kramdown output is mangled! 😵‍💫

> I don't see this functionality as something which would be added directly to core.

I understand. Given how it easy it was to write that dumb unwrapping plugin, that's fair enough. 

> I'm not quite sure I follow the part where all of the newlines need to be replaced.

Me neither 😂
